src/sdk/namespace_fs.js

```diff
@@ -682,44 +682,83 @@ class NamespaceFS {
         const fs_account_config = this.set_cur_fs_account_config(object_sdk);
         await this._load_bucket(params, fs_account_config);
 
+        const can_use_temp_file = Boolean(fs_account_config.backend === 'GPFS' && nb_native().fs.gpfs);
         const file_path = this._get_file_path(params);
-        // Uploading to a temp file then we will rename it. 
-        const upload_id = uuidv4();
-        const upload_path = path.join(this.bucket_path, this.get_bucket_tmpdir(), 'uploads', upload_id);
-        let fs_xattr = to_fs_xattr(params.xattr);
-        // dbg.log0('NamespaceFS.upload_object:', upload_path, '->', file_path);
         try {
-            await this._check_path_in_bucket_boundaries(fs_account_config, file_path);
-            await this._make_path_dirs(upload_path, fs_account_config);
-            if (params.copy_source) {
-                const source_file_path = path.join(this.bucket_path, params.copy_source.key);
-                await this._check_path_in_bucket_boundaries(fs_account_config, source_file_path);
-                await this._fail_if_archived_or_sparse_file(fs_account_config, source_file_path);
-                try {
-                    const same_inode = await this._is_same_inode(fs_account_config, source_file_path, file_path, fs_xattr);
-                    if (same_inode) return same_inode;
-                    // Doing a hard link.
-                    await nb_native().fs.link(fs_account_config, source_file_path, upload_path);
-                } catch (e) {
-                    dbg.warn('NamespaceFS: COPY using link failed with:', e);
-                    fs_xattr = await this._copy_stream(source_file_path, upload_path, fs_account_config, fs_xattr);
+            if (can_use_temp_file) {
+                const open_mode = 'wt';
+                // Uploading to a temp file then we will rename it.
+                let fs_xattr = to_fs_xattr(params.xattr);
+                // dbg.log0('NamespaceFS.upload_object:', this.upload_path, '->', file_path);
+                if (params.copy_source) {
+                    const upload_id = uuidv4();
+                    const upload_path = path.join(this.bucket_path, this.get_bucket_tmpdir(), 'uploads', upload_id);
+                    const source_file_path = path.join(this.bucket_path, params.copy_source.key);
+                    await this._check_path_in_bucket_boundaries(fs_account_config, file_path);
```

Comment:
This can be pushed up outside of `if (can_use_temp_file)`

Reply:
This has to be called for all of our cases, both tmpfile and copy_source, so should end up in the try{} level.

Reply:
done